feat(martin-server):  add support for basic CORS configuration (#1815)
urbica/martin: docs/src/config-file.md

# Enable or disable Martin web UI. At the moment, only allows `enable-for-all` which enables the web UI for all connections. This may be undesirable in a production environment. [default: disable]
web_ui: disable

# CORS Configuration

> # Can also be disabled via `cors: false`

Reviewer: need to document what does `true` means, and what happens if not set at all (default).  Note that default here and in sub-items is a bit confusing, i.e. does `cors: {}` equal to `cors: ~` ?

Author: good point - i'll expand on that and double check what happens. might even add some tests
never seen  `~` or `{}` in the wild though.

Author: Right now, both `cors: {}` and `cors: ~` seem to enable CORS using our defaults.

Author: As it stands right now, it'll always fall back to the defaults. 

- no `cors` option at all -> Enable using defaults
- `cors: true` -> Enable using defaults
- only configuring a subset and omitting certain properties-> use defaults for properties that are not explicitly set. 

 this...
```yaml
cors:
  max_age: 3600
```
would translate to this:
```yaml
cors:
  origin: ["*"]
  max_age: 3600
```
The only way to disable CORS right now, is to explicitly set `cors: false`. 

I guess `cors: ~` should also disable it? What would you prefer

Reviewer: No, null is the same as skipped, so it should be the default

cors:

>   # Sets the `Access-Control-Allow-Origin` header [default: *]

Reviewer: just to double check -- does the default of `*` mean exactly the same as the current default behavior - i.e. allow all?  I.e. currently it "will use the requests `ORIGIN` header" ?

Author: yep, "will use the requests ORIGIN header", is the current default .

[docs.rs](https://docs.rs/actix-cors/latest/actix_cors/struct.Cors.html#method.allowed_origin) 
`send_wildcard()` is disabled by default in `actix_cors`. 

> When all origins are allowed and send_wildcard is set, * will be sent in the Access-Control-Allow-Origin response header. If send_wildcard is not set, the client’s Origin request header will be echoed back in the Access-Control-Allow-Origin response header.

Author: To expand on this a bit, this is the current CORS initialization on `main`:
```rust
    let cors_middleware = Cors::default()
            .allow_any_origin()
            .allowed_methods(vec!["GET"]);
```

- `default()`: sets restrictive defaults
*No* allowed origins, methods, request headers or exposed headers. Credentials
not supported. No max age (will use browser's default).
- `allow_any_origin()`: allows any origin, either by returning `*` in the header or by just returning the requests `ORIGIN` (this is the default in actix_cors as `send_wildcard()` is NOT enabled). .
- `allowed_methods(vec!["GET"])` sets `Access-Control-Allow-Methods`, only used in OPTIONS/preflight requests. 

This PR keeps the exact same defaults for now, but allows disabling CORS altogether and dynamically setting `ACCESS_CONTROL_ALLOW_ORIGIN` and `ACCESS_CONTROL_MAX_AGE`. 

If the defaults were to change at some point (this would be a breaking change), we could think about making it a bit more restrictive by default.

Maybe enable CORS by default, but require users to set specific origins. This would also ensure that an informed decision is made when opting out of "security" by allowing any origin or disabling it.

> # CORS configuration
> cors:
>   # enable/disable CORS [default: true]
>   enable: true
>   # sets the Access-Control-Allow-Origin header [default: *]
>   # '*' will use the requests ORIGIN header
>   origin:
>     - https://example.org
>   # sets Access-Control-Max-Age Header. Remove to use default. [default: None]
>   max_age: 3600

Reviewer 2: Mostly, this is great work ❤️ 

I don't quite like the `enable: true` part of the config.
If CORS is entirely unconfigured, the rest of the options don't quite make sense.

What about below?

```suggestion
# CORS configuration
#
# Can also be disabled via `cors: false`
cors:
  # Sets the Access-Control-Allow-Origin header [default: *]
  # '*' will use the requests ORIGIN header
  origin:
    - https://example.org
  # Sets Access-Control-Max-Age Header. [default: null]
  # null means not seting the header
  max_age: 3600
```

(And changing the `CorsConfig` to be an enum and with `#[serde(transparent)]` on the variant with configuration)

# Enable or disable Martin web UI. At the moment, only allows `enable-for-all` which enables the web UI for all connections. This may be undesirable in a production environment. [default: disable]
web_ui: disable

# CORS Configuration

> # Defaults to `cors: true`, which allows all origins.

Reviewer 2: Formatting nit

```suggestion
#
# Defaults to `cors: true`, which allows all origins.
```

> # Can be completely disabled via `cors: false`

Reviewer 2: ```suggestion
# Sending/acting on cors headers be completely disabled via `cors: false`
```
